### ISSUE TITLE
Brute force nearest neighbor search operator scaling with num-threads-per-search and filter

### ISSUE THREAD

**Issue by jobergum** [vespa-engine/vespa#18900]
The brute force [nearest neighbor search operator](https://docs.vespa.ai/en/nearest-neighbor-search.html) experiences lock contention when used in combination with higher num-threads-per-search than 1 and when there are other filter constraints in the overall query tree. 

For example given a tensor embedding field without HNSW indexing which is used for brute force nearest neighbor search. Note that this is not relevant for use cases using **approximate** [nearest neighbor search with HNSW](https://docs.vespa.ai/en/approximate-nn-hnsw.html).  

Note that using _num-threads-per-search_ only tunes down the default number (persearch) and is configured in the [services.xml.](https://docs.vespa.ai/en/reference/services-content.html#requestthreads-persearch) 

```
schema foo 
  document foo { 
    field embedding type tensor<float>(x[512]) {
      indexing: attribute
      attribute {
        distance-metric:euclidean
       }
    }
    field gid type long {
       indexing: attribute
       attribute:fast-search
       rank:filter
    }
 }
  rank-profile euclidean-t24 {
    num-threads-per-search:24
    first-phase {
         expression: closeness(field,embedding)
    }
 }
}
```
A brute force NN query like 

```
{
 "yql": "select * from * where [{\"targetNumHits\": 10}]nearestNeighbor(embedding, query_embedding);",
  "ranking.profile": "euclidean-t24",
   "ranking.features.query(query_embedding)": [......]
}
```
Would work well and as expected using 24  threads to brute force compute nn, however when used in combination with a filter to limit the brute force nn search like filtering on gid=4:

```
{
 "yql": "select * from * where [{\"targetNumHits\": 10}]nearestNeighbor(embedding, query_embedding) and gid=4;",
  "ranking.profile": "euclidean-t4",
   "ranking.features.query(query_embedding)": [......]
}
```
It would suffer from lock contention and if the filter criteria matches a lot of documents (e.g 50%) (but less than all) the run time performance would be worse than just brute forcing 100% . This scaling problem with many threads per search  is addressed in https://github.com/vespa-engine/vespa/pull/18893

**Alternative work arounds**
It's possible to drop using the nearest neighbor search operator and instead express the distance function using tensor math replacing the _closeness(field,_ embedding)_ ranking feature which is calculated by the nearestNeighbor search query operator with a tensor expression calculating the euclidean distance. Note that we want to rank documents with small distance so we use 1/(1 + euclideanDistanceFormula) as our ranking score. This is the same behaviour as the _closeness(field,embedding)_ feature does. 

```
 rank-profile euclidean-tensor-ranking {
    num-threads-per-search:24
    first-phase {
         expression: 1/(1 + sqrt(sum(map(query(query_embedding) - attribute(embedding), f(x)(x * x)))))
    }
 }
 rank-profile tensor-dotproduct {
    num-threads-per-search:24
    first-phase {
         expression: sum(query(query_embedding) * attribute(embedding))
    }
 }
```

Where the query retriever logic would be just searching for gid. This scales well without the lock contention problems described above, but will be slower than brute force nearestNeighbor once a release is made which has https://github.com/vespa-engine/vespa/pull/18893. The reason it is slower is that the above euclidean distance calculation would not be recognized and HW accelerated. The example in _tensor-dotproduct_ however would be accelerated, but generally not all possible ways of writing tensor expressions are recognized and optimized with HW vector instructions. 

Query example using tensor expression ranking:

```
{
 "yql": "select * from * where gid=4;",
  "ranking.profile": "euclidean-tensor-ranking",
   "ranking.features.query(query_embedding)": [......]
}
```

**Comment by jobergum** [comment]
Also https://github.com/vespa-engine/vespa/pull/18898 will make the brute force (approximate:false) nearestNeigbor search operator faster, especially for large number of vectors per node.

**Comment by jobergum** [comment]
A fix for the above problems is included in 7.457.52 and above.
